### PR TITLE
DatasourceVariables: Update query editor when switching datasources from picker

### PR DESCRIPTION
**Pull request by kaydelaney**
Fixes an issue (described here: #52737) where choosing a different instance of a datasource from a datasource variable picker wouldn't refresh the query editor as you might expect.

**Which issue(s) does this PR fix?**:
Closes #52737 #63904

### CODE REVIEW

**Review by axelavargas — COMMENTED** [review]
Hey, @kaydelaney :), I gave it a spin and the changes implemented here fixed the problem 🥳 .

Unfortunately, I can't review the code properly as I don't have enough context of how the `QueryEditorRow` works. I am adding @torkelo as he was working recently on this file when fixing this other issue https://github.com/grafana/grafana/pull/61927/files 

Also, worth double-checking if this PR fixes this issue that was recently reported: https://github.com/grafana/grafana/issues/63904

**Review by torkelo — CHANGES_REQUESTED** [review]
I am not sure about this change, it feels messy and hard to grasp with different logic everywhere (different logic in getDataSourceIdentifier and what is used in comparisons).

I propose this change, to simplify and unify the logic 

```patch
diff --git a/public/app/features/query/components/QueryEditorRow.tsx b/public/app/features/query/components/QueryEditorRow.tsx
index ea1605dd3f..2dc2ecd4e8 100644
--- a/public/app/features/query/components/QueryEditorRow.tsx
+++ b/public/app/features/query/components/QueryEditorRow.tsx
@@ -77,6 +77,7 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
   element: HTMLElement | null = null;
   angularScope: AngularQueryComponentScope<TQuery> | null = null;
   angularQueryEditor: AngularComponent | null = null;
+  dataSourceSrv = getDataSourceSrv();
   id = '';
 
   state: State<TQuery> = {
@@ -136,25 +137,32 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
     };
   }
 
-  getQueryDataSourceIdentifier(): string | null | undefined {
-    const { query, dataSource: dsSettings } = this.props;
-    return query.datasource?.uid ?? dsSettings.uid;
+  /**
+   * When datasource variables are used the query.datasource.uid property is a string variable expression
+   * DataSourceSettings.uid can also be this variable expression.
+   * This funtion always returns the current interpolated datasource uid.
+   */
+  getInterpolatedDataSourceUID(): string | null | undefined {
+    if (this.props.query.datasource) {
+      return this.dataSourceSrv.getInstanceSettings(this.props.query.datasource)?.rawRef?.uid;
+    }
+
+    return this.props.dataSource.rawRef?.uid;
   }
 
   async loadDatasource() {
-    const dataSourceSrv = getDataSourceSrv();
     let datasource: DataSourceApi;
-    const dataSourceIdentifier = this.getQueryDataSourceIdentifier();
+    const interpolatedUID = this.getInterpolatedDataSourceUID();
 
     try {
-      datasource = await dataSourceSrv.get(dataSourceIdentifier);
+      datasource = await this.dataSourceSrv.get(interpolatedUID);
     } catch (error) {
-      datasource = await dataSourceSrv.get();
+      datasource = await this.dataSourceSrv.get();
     }
 
     this.setState({
       datasource: datasource as unknown as DataSourceApi<TQuery>,
-      loadedDataSourceIdentifier: dataSourceIdentifier,
+      loadedDataSourceIdentifier: interpolatedUID,
       hasTextEditMode: has(datasource, 'components.QueryCtrl.prototype.toggleEditorMode'),
     });
   }
@@ -182,7 +190,7 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
     }
 
     // check if we need to load another datasource
-    if (datasource && loadedDataSourceIdentifier !== this.getQueryDataSourceIdentifier()) {
+    if (datasource && loadedDataSourceIdentifier !== this.getInterpolatedDataSourceUID()) {
       if (this.angularQueryEditor) {
         this.angularQueryEditor.destroy();
         this.angularQueryEditor = null;
@@ -243,7 +251,7 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
   isWaitingForDatasourceToLoad(): boolean {
     // if we not yet have loaded the datasource in state the
     // ds in props and the ds in state will have different values.
-    return this.props.dataSource.uid !== this.state.loadedDataSourceIdentifier;
+    return this.getInterpolatedDataSourceUID() !== this.state.loadedDataSourceIdentifier;
   }
 
   renderPluginEditor = () => {

```

**Comment by kaydelaney** [comment]
@torkelo The changes you suggested seemed to break switching between non-variable datasources (since they don't have `rawRef`s it would just fall back to the default datasource), so I changed `getInterpolatedDataSourceUID` to return the `uid` if `rawRef` doesn't exist

**Comment by torkelo** [comment]
Dam, rawRef is undefined sometimes? That makes zero sense , just pointlessly increasing complexity. Can we always return it? Or switch it so that instead of returning rawRef we return the sourceUid (which is the string containing variable expression) and have datasourceSettings.uid always be the real uid, and then remove rawRef ?

**Review by torkelo — APPROVED** [review]

**Comment by torkelo** [comment]
but let's merge this, and revisit changing rawRef later :)

File: public/app/features/query/components/QueryEditorRow.tsx [hunk 1]
```diff
@@ -243,7 +252,7 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
   isWaitingForDatasourceToLoad(): boolean {
     // if we not yet have loaded the datasource in state the
     // ds in props and the ds in state will have different values.
-    return this.props.dataSource.uid !== this.state.loadedDataSourceIdentifier;
+    return this.getInterpolatedDataSourceUID() !== this.state.loadedDataSourceIdentifier;
```
Comment (ivanortegaalba) [inline public/app/features/query/components/QueryEditorRow.tsx:255]: This is solving the issue of not displaying the editor for certain scenarios #64011. 

But I found it confusing that `this.state.loadedDataSourceIdentifier` contains the value we're trying to load, even if the datasource doesn't exist. When it doesn't exist, the loaded data source will be the default one, but the value in this state property will be `this.getInterpolatedDataSourceUID()`.

This is very confusing, so maybe we can just change the name of the state property to `queriedDataSource`. WDYT?

PS: _I spent a considerable amount of time understanding why the state value is different than the one rendering the editor_

**Review by ivanortegaalba — APPROVED** [review]
This solves the issue that the PR https://github.com/grafana/grafana/pull/64011 was trying to solve but in a better way and without regressions. Thanks, @torkelo to point me here.

This fixes #63904

File: public/app/features/query/components/QueryEditorRow.tsx [hunk 2]
```diff
@@ -136,25 +137,33 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
...
     let datasource: DataSourceApi;
-    const dataSourceIdentifier = this.getQueryDataSourceIdentifier();
+    const interpolatedUID = this.getInterpolatedDataSourceUID();
 
     try {
-      datasource = await dataSourceSrv.get(dataSourceIdentifier);
+      datasource = await this.dataSourceSrv.get(interpolatedUID);
     } catch (error) {
-      datasource = await dataSourceSrv.get();
+      datasource = await this.dataSourceSrv.get();
```
Comment (ivanortegaalba) [inline public/app/features/query/components/QueryEditorRow.tsx:161]:
```suggestion
      // If the DS doesn't exist, it fails. When getting with no args, it returns the default DS.
      datasource = await this.dataSourceSrv.get();
```

File: public/app/features/query/components/QueryEditorRow.tsx (hunk 1 above)
Comment (ivanortegaalba) [inline public/app/features/query/components/QueryEditorRow.tsx:255]:
```suggestion
    return this.getInterpolatedDataSourceUID() !== this.state.queriedDataSource;
```

File: public/app/features/query/components/QueryEditorRow.tsx [hunk 3]
```diff
@@ -136,25 +137,33 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
...
     }
 
     this.setState({
       datasource: datasource as unknown as DataSourceApi<TQuery>,
-      loadedDataSourceIdentifier: dataSourceIdentifier,
+      loadedDataSourceIdentifier: interpolatedUID,
```
Comment (ivanortegaalba) [inline public/app/features/query/components/QueryEditorRow.tsx:166]:
```suggestion
      queriedDataSource: interpolatedUID,
```

File: public/app/features/query/components/QueryEditorRow.tsx (hunk 1 above)
Comment (torkelo) [reply public/app/features/query/components/QueryEditorRow.tsx:255]: @ivanortegaalba hm.. if you get the default data source when asking for a one that does not exist, maybe we should change that?

Comment (ivanortegaalba) [reply public/app/features/query/components/QueryEditorRow.tsx:255]: Just to be clear, when we call `this.dataSourceSrv.get(interpolatedUID)` it fails. So the logic for dataSourceSrv is correct. 
We get the default datasource [here](https://github.com/grafana/grafana/pull/62617/files#diff-d941f1ab9516ae305170c05387d11f8cb39972e8a67e001c1840b7f6985710a1R156-R161) once the request has failed, we just assign the default one. But the `loadedDataSourceIdentifier` won't have the default datasource UID. This was the main confusion

**Comment by gillesdouaire** [comment]
Hello! 
This is tagged to be included in 9.4.5, but it's not in 9.4.7. What are your plans - just want to make sure this is not forgotten. 

Thanks!

**Pull request merged**